arc-script/arc-script-core/src/compiler/arcon/lower/hir/mod.rs

```diff
@@ -1,6 +1,5 @@
 pub(crate) mod lowerings {
     pub(crate) mod structs;
-    pub(crate) mod tasks;
```

Comment:
In the ideal case I do not want to have `timeout`/`timer`/`trigger`. The reason I did not add it yet is because the timer-events have a bit of a different semantics compared to message-events.

```
task Delay() (In(~i32)) -> (Out(~i32)) {
    port Timer(~i32);
    on {
        In(event) => emit Timer(event) after 1h,
        Timer(event) => emit Out(event)
    }
}
```

I would expect the above code to delay every event by 1h. In Arcon, there can only be one timer scheduled at any given point in time. This would mean the internal-port message-queue can hold up to one event. If a new message arrives the old is discarded.

The erlang-style `after` is an interesting idea, I need to think about what you can build with it 🤔

Reply:
The second reason for not adding it yet is because I found it to be a bit tricky from a type system perspective.

This code:
```
task Delay() (In(~i32)) -> (Out(~i32)) {
    port Timer(~i32);
    on {
        In(event) => emit Timer(event) after 1h,
        Timer(event) => emit Out(event)
    }
}
```

Internally it desugars to:

```
enum Source { In(~i32), Timer(~i32) }
enum Sink { Out(~i32), Timer(~i32) }
task Delay() Source -> Sink {
    on {
        In(event) => emit Timer(event) after 1h,
        Timer(event) => emit Out(event)
    }
}
```
You need to infer somehow that it is:
```
enum Source { In(~i32), Timer(~i32) }
enum Sink { Out(~i32), Timer(~i32) }
task Delay() Source -> Sink {
    on {
        Source::In(event) => emit Sink::Timer(event) after 1h,
        Source::Timer(event) => emit Sink::Out(event)
    }
}
```
this is tough because you could write
```
enum Source { In(~i32), Timer(~i32) }
enum Sink { Out(~i32), Timer(~i32) }
task Delay() Source -> Sink {
    on {
        event => foo(event),
    }
}
```
For it to work I think you need to treat enums as structural types instead of nominal.

Reply:
You could provide you own timer mechanism in arcorn, you track the scheduled event time for all events and set the arcon timer to the earliest time. When it triggers you figure out which of the events should trigger.

> The erlang-style after is an interesting idea, I need to think about what you can build with it
You typically use it for detecting when you have been idle for too long.

In arc-script you could use it to put a cap on the latency due to windowing, i.e. you either trigger a window operation when you have received N events or no events have arrived for S seconds.

Reply:
> You could provide you own timer mechanism in arcorn, you track the scheduled event time for all events and set the arcon timer to the earliest time. When it triggers you figure out which of the events should trigger.

This is very interesting, I did not think of it but it should definitely be possible to implement, maybe even at the "Arcorn level" 

> n arc-script you could use it to put a cap on the latency due to windowing, i.e. you either trigger a window operation when you have received N events or no events have arrived for S seconds.

Excellent idea

Reply:
Is there any other mechanism we could borrow from Erlang?

Reply:
Something we should not borrow is selective receive, that would not play well with arcon :)

Reply:
I borrowed the `or`, `bor`, `band`, `bor` operators

Reply:
I wonder if we could borrow atoms, but maybe it does not make sense when you have types

Reply:
> I wonder if we could borrow atoms, but maybe it does not make sense when you have types

As we have Rust-style enums I don't think it is worth it.